### PR TITLE
Implement `return_indices` option to `F.max_pooling_2d`

### PR DESCRIPTION
**Pull request by kmaehashi**
Refs https://github.com/chainer/chainer/pull/4421#issuecomment-395377917

### CODE REVIEW

**Comment by kmaehashi** [comment]
Thanks for the comment. I fixed to test validity of indices.
(Like as we test validity of outputs here: https://github.com/kmaehashi/chainer/blob/180a61777e474b3dab6e7c0f3881e935c15771df/tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py#L72-L87)

File: tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py [hunk 1]
```diff
@@ -224,4 +224,46 @@ def test_call_cudnn_backward(self):
...
+        out, indices = functions.max_pooling_2d(
+            x, 2, cover_all=False, return_indices=True)
+        assert isinstance(out, chainer.Variable)
+        assert isinstance(out.array, type(x))
+        assert isinstance(indices, type(x))
+        assert indices.shape == out.array.shape
+
+        # Calculate expected indices.
+        expect = numpy.zeros(indices.shape, dtype=indices.dtype)
+        for k in six.moves.range(2):
```
[the remark this answers is not in the document]
Comment (kmaehashi) [reply tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py]: Makes sense. How about `i` instead?

[the remark this answers is not in the document]
Comment (hvy) [reply tests/chainer_tests/functions_tests/pooling_tests/test_max_pooling_2d.py]: Sounds good to me!

File: chainer/functions/pooling/max_pooling_2d.py [hunk 2]
```diff
@@ -356,9 +357,25 @@ def max_pooling_2d(x, ksize, stride=None, pad=0, cover_all=True):
...
+            ``return_indices`` is set to ``True``, as cuDNN does not return
+            indices information.
 
     Returns:
-        ~chainer.Variable: Output variable.
+        ~chainer.Variable or tuple:
+            When ``return_indices`` is ``False`` (default), returns the output
+            variable. Otherwise returnes the tuple of the output variable and
+            pooling indices (`ndarray`). Pooling indices will be on the same
+            device as the input.
```
[the remark this answers is not in the document]
Comment (kmaehashi) [reply chainer/functions/pooling/max_pooling_2d.py]: Fixed.

File: chainer/functions/pooling/max_pooling_2d.py [hunk 3]
```diff
@@ -356,9 +357,26 @@ def max_pooling_2d(x, ksize, stride=None, pad=0, cover_all=True):
...
+            Note that cuDNN will not be used for this function if
+            ``return_indices`` is set to ``True``, as cuDNN does not return
+            indices information.
 
     Returns:
-        ~chainer.Variable: Output variable.
+        ~chainer.Variable or tuple:
+            When ``return_indices`` is ``False`` (default), returns the output
+            variable.
+            When ``True``, returnes the tuple of the output variable and
```
Comment (hvy) [inline chainer/functions/pooling/max_pooling_2d.py]: Could you fix the typo `returnes` -> `returns` ?

Comment (kmaehashi) [reply chainer/functions/pooling/max_pooling_2d.py]: Thanks, fxed.

**Comment by hvy** [comment]
Thanks for the quick fix, LGTM! I'll merge it once CI passes.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
LGTM!

**Pull request merged**